actionpack/lib/action_dispatch/routing/routes_proxy.rb

```diff
@@ -58,6 +58,9 @@ def #{method}(*args)
       def merge_script_names(previous_script_name, new_script_name)
         return new_script_name unless previous_script_name
 
+        # This check to resolve: https://github.com/rails/rails/issues/31476
+        return previous_script_name if Rails.application.config.relative_url_root == previous_script_name && new_script_name == '/'
```

Comment:
We can't use `Rails.application` here. `relative_url_root` is inside the `RouteSet#relative_url_root`

Reply:
relative_url_root inside of RouteSet#relative_url_root is incorrect AND is caused to always return "" by recent code changes. The below is from my description of this from the issue of this bug:

From investigation, this issue sprung from the following logic no longer being used that I can see. The first of the two sections no longer being valid would be: https://github.com/rails/rails/blob/v5.2.0.beta2/actionpack/lib/action_dispatch/routing/route_set.rb#L347

The issue is that the "config" being loaded here is the default RAILS configuration rather than the local applications overridden settings. So while my sample application has set config.relative_url_root in config/environments/development.rb, that value doesn't exist when the above code runs. As the default RAILS configuration does look for a RAILS_RELATIVE_URL_ROOT environment variable, one can get that set correctly in my sample apps case with: RAILS_RELATIVE_URL_ROOT="/v2" rails s

That then brings me to the second issue that I have less of an understanding regarding. Six months ago the following line was added that forces an engine route to claim it isn't deployed into a subdirectory: https://github.com/rails/rails/blob/v5.2.0.beta2/actionpack/lib/action_dispatch/routing/mapper.rb#L670

Because of that setting, the following code will only ever return the empty "" value rather than the correct "subdirectory" path (if that configuration was working correctly): https://github.com/rails/rails/blob/v5.2.0.beta2/actionpack/lib/action_dispatch/routing/route_set.rb#L782

Fixing this would require changing how recent logic has been implemented. I can attempt to do this if this is the preferred approach but the margin for logic errors greatly increases as I'm unsure the reasoning as to why others broken this functionality? Let me know!

Reply:
> So while my sample application has set config.relative_url_root in config/environments/development.rb, that value doesn't exist when the above code runs.

This is the issue we should be fixing. It doesn't make sense to us to have that code there if when it is executed the config is not set yet.

Reply:
@rafaelfranca : I resolved the more minor issues and have this using the setting from the relative_url_root from RouteSet. However, I'm unsure how to get the value set in the Rails application config.relative_url_root. The `railties/lib/engine.rb` is what initializes the `route_set.rb` config but has no hook into anything outside of the engine itself at that point.

I can work around this by setting the environment variable in my application that the engine's default configuration can pick up. If you have suggestions on how I should get the Rails application configuration at that point, I would need the help. :)

Otherwise, I can also update the documents at: http://guides.rubyonrails.org/configuring.html#deploy-to-a-subdirectory-relative-url-root to encourage setting the default Environment variable over a hard-coded path and/or having engines allow for themselves to configured with a "relative_url_root" setting inside of their own local configuration setup?

Thanks!